[GEOT-5738] URLs.urlToFile does not percent-decode URLs with no slash after "file:"

johngschulz opened the pull request:
https://osgeo-org.atlassian.net/browse/GEOT-5738

tbarsballe commented on modules/library/metadata/src/main/java/org/geotools/util/URLs.java:
```diff
@@ -186,7 +186,12 @@ public static File urlToFile(URL url) {
+            try {
+                path2 = URLDecoder.decode(path2, "UTF-8");
+            } catch (UnsupportedEncodingException e) {
+                throw new RuntimeException("Could not decode the URL to UTF-8 format", e);
```
Throwing a runtime exception here seems a bit severe.
However, I see that that is what the method is already doing for a similar failure earlier, and that by the WC3 spec URL encoding is assumed to be UTF-8.

I am still a bit concerned this could cause problems, but it does seem to be OK.

tbarsballe commented on modules/library/metadata/src/test/java/org/geotools/util/URLsTest.java:
```diff
@@ -112,6 +112,7 @@ public void testUrlToFile() throws Exception {
         URL url = new URL("file", "////oehhwsfs09", "/some/path/on/the/server/filename.nds");
         File windowsShareFile = URLs.urlToFile(url);
         assertNotNull(windowsShareFile);
+        assertEquals("file café", URLs.urlToFile(new URL("file:file%20caf%C3%A9")).toString());
```
Test case is a bit sparse, but adequate. For completeness, testing all of the following might be better, bit it is OK as-is:

`file:file%20caf%C3%A9"`
`file:/file%20caf%C3%A9"`
`file://file%20caf%C3%A9"`

bencaradocdavies replied:
@johngschulz please add the extra tests suggested by @tbarsballe. We will thank you in the future.

johngschulz replied:
@bencaradocdavies I've added the additional test, when converting urlToFile using `file:/` the single slash remains in place.  I am unsure if this is an expected behavior or not.

johngschulz commented:
If I understood correctly - I've changed the exception to an IllegalArgumentException and the associated message, changed the "UTF-8" string to java.nio.charset.StandardCharsets.UTF_8.name(), and added the additional tests.

I've added a note about how the single slash is handled. Let me know if the changes look appropriate. Thanks

bencaradocdavies commented:
@johngschulz thanks very much. Those changes look good. For the single-slash case your test is just documenting the existing behaviour. All these cases are to various degrees dodgy non-conforming uses. I cherry-picked your change locally (so I could remove one extra blank line) and ran a full GeoTools build + full GeoServer build + app-schema online test to make sure integration works. All good, so pushed to master.

johngschulz commented:
Thanks cheers!
